### PR TITLE
new data source: `azurerm_kubernetes_cluster_snapshot`

### PR DESCRIPTION
**Pull request by ms-henglu**
```
=== RUN   TestAccKubernetesCluster_snapshotId
=== PAUSE TestAccKubernetesCluster_snapshotId
=== CONT  TestAccKubernetesCluster_snapshotId
--- PASS: TestAccKubernetesCluster_snapshotId (1377.46s)
PASS

=== RUN   TestAccDataSourceKubernetesClusterSnapshot_basic
=== PAUSE TestAccDataSourceKubernetesClusterSnapshot_basic
=== CONT  TestAccDataSourceKubernetesClusterSnapshot_basic
--- PASS: TestAccDataSourceKubernetesClusterSnapshot_basic (656.52s)
PASS

```

### CODE REVIEW

File: internal/services/containers/kubernetes_cluster_resource.go [hunk 1]
```diff
@@ -1142,6 +1143,13 @@ func resourceKubernetesCluster() *pluginsdk.Resource {
 				}, false),
 			},
 
+			"snapshot_id": {
```
Comment (katbyte) [inline internal/services/containers/kubernetes_cluster_resource.go:1146]: should this be
```suggestion
			"create_from_snapshot_id": {
```

File: internal/services/containers/kubernetes_cluster_snapshot_data_source.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+			resp, err := client.Get(ctx, id)
+			if err != nil {
+				if response.WasNotFound(resp.HttpResponse) {
+					return fmt.Errorf("%s was not found", id)
+				}
+				return fmt.Errorf("retrieving %s: %v", id, err)
+			}
+
+			state.Name = id.ManagedClusterSnapshotName
```
Comment (katbyte) [inline internal/services/containers/kubernetes_cluster_snapshot_data_source.go:88]: why are we doing this?

File: internal/services/containers/kubernetes_cluster_snapshot_data_source_test.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+							CreationData: &managedclustersnapshots.CreationData{
+								SourceResourceId: utils.String(clusterId.ID()),
+							},
+						},
+					}
+					_, err = client.CreateOrUpdate(ctx, id, snapshot)
+					if err != nil {
+						return fmt.Errorf("creating %s: %+v", id, err)
+					}
+					return nil
```
Comment (katbyte) [inline internal/services/containers/kubernetes_cluster_snapshot_data_source_test.go:47]: why are we doing this in the test?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the pr @ms-henglu ,

however i'm really not sure what this datasource is for or doing, could you please explain what it is doing and why it exists?

**Comment by ms-henglu** [comment]
Hi @katbyte ,

Thanks for the code review!

The use case is that user can create a snapshot for a aks cluster, then restore it to a new aks cluster. But we couldn't make the snapshot as an azurerm resource, because when we do that, the snapshot is depending on the source cluster, then users couldn't remove the source cluster without destroying the snapshot in terraform, that's why we need to create the snapshot in the test step.

And this data source is used to help user construct the snapshot ID.

A similar resource is already released: https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/containers/kubernetes_node_pool_snapshot_data_source.go

File: internal/services/containers/kubernetes_cluster_resource.go (hunk 1 above)
Comment (ms-henglu) [reply internal/services/containers/kubernetes_cluster_resource.go:1146]: I think we should align with the similar feature in node pool: https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/containers/kubernetes_cluster_node_pool_resource.go#L298

File: internal/services/containers/kubernetes_cluster_other_resource_test.go [hunk 4]
```diff
@@ -791,6 +797,65 @@ func TestAccKubernetesCluster_azureMonitorKubernetesMetrics(t *testing.T) {
...
+				data.CheckWithClientForResource(func(ctx context.Context, clients *clients.Client, state *terraform.InstanceState) error {
+					client := clients.Containers.ManagedClusterSnapshotClient
+					clusterId, err := managedclusters.ParseManagedClusterID(state.ID)
+					if err != nil {
+						return err
+					}
+					id := managedclustersnapshots.NewManagedClusterSnapshotID(clusterId.SubscriptionId, clusterId.ResourceGroupName, data.RandomString)
+					_, err = client.Delete(ctx, id)
+					if err != nil {
+						return fmt.Errorf("creating %s: %+v", id, err)
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_other_resource_test.go:850]:
```suggestion
						return fmt.Errorf("deleting %s: %+v", id, err)
```

File: internal/services/containers/kubernetes_cluster_snapshot_data_source_test.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+					if err != nil {
+						return err
+					}
+					id := managedclustersnapshots.NewManagedClusterSnapshotID(clusterId.SubscriptionId, clusterId.ResourceGroupName, data.RandomString)
+					snapshot := managedclustersnapshots.ManagedClusterSnapshot{
+						Location: data.Locations.Primary,
+						Properties: &managedclustersnapshots.ManagedClusterSnapshotProperties{
+							SnapshotType: utils.ToPtr(managedclustersnapshots.SnapshotTypeManagedCluster),
+							CreationData: &managedclustersnapshots.CreationData{
+								SourceResourceId: utils.String(clusterId.ID()),
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_snapshot_data_source_test.go:39]:
```suggestion
								SourceResourceId: pointer.To(clusterId.ID()),
```

File: internal/services/containers/kubernetes_cluster_snapshot_data_source_test.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+				data.CheckWithClientForResource(func(ctx context.Context, clients *clients.Client, state *terraform.InstanceState) error {
+					client := clients.Containers.ManagedClusterSnapshotClient
+					clusterId, err := managedclusters.ParseManagedClusterID(state.ID)
+					if err != nil {
+						return err
+					}
+					id := managedclustersnapshots.NewManagedClusterSnapshotID(clusterId.SubscriptionId, clusterId.ResourceGroupName, data.RandomString)
+					_, err = client.Delete(ctx, id)
+					if err != nil {
+						return fmt.Errorf("creating %s: %+v", id, err)
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_snapshot_data_source_test.go:69]:
```suggestion
						return fmt.Errorf("deleting %s: %+v", id, err)
```

File: internal/services/containers/registration.go [hunk 7]
```diff
@@ -58,6 +58,7 @@ func (r Registration) SupportedResources() map[string]*pluginsdk.Resource {
 func (r Registration) DataSources() []sdk.DataSource {
 	dataSources := []sdk.DataSource{
 		KubernetesNodePoolSnapshotDataSource{},
+		KubernetesClusterSnapshotDataSource{},
```
Comment (stephybun) [inline internal/services/containers/registration.go:61]: Alphabetical ordering - this should be above the node pool data source

**Review by stephybun — CHANGES_REQUESTED** [review]

File: website/docs/d/kubernetes_cluster_snapshot.html.markdown [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+* `name` - The name of the Kubernetes Cluster Snapshot.
+
+* `resource_group_name` - The name of the Resource Group in which the Kubernetes Cluster Snapshot exists.
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `id` - The ID of the Kubernetes Node Pool Snapshot.
```
Comment (stephybun) [inline website/docs/d/kubernetes_cluster_snapshot.html.markdown:34]:
```suggestion
* `id` - The ID of the Kubernetes Cluster Snapshot.
```

File: website/docs/d/kubernetes_cluster_snapshot.html.markdown [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+---
+
+# Data Source: azurerm_kubernetes_cluster_snapshot
+
+Use this data source to access information about an existing Kubernetes Cluster Snapshot.
+
+## Example Usage
+
+```hcl
+data "azurerm_kubernetes_node_pool_snapshot" "example" {
```
Comment (stephybun) [inline website/docs/d/kubernetes_cluster_snapshot.html.markdown:16]:
```suggestion
data "azurerm_kubernetes_cluster_snapshot" "example" {
```

File: website/docs/r/kubernetes_cluster.html.markdown [hunk 10]
```diff
@@ -235,6 +235,8 @@ resource "azurerm_kubernetes_cluster" "example" {
 
 -> **Note:** Whilst the AKS API previously supported the `Paid` SKU - the AKS API introduced a breaking change in API Version `2023-02-01` (used in v3.51.0 and later) where the value `Paid` must now be set to `Standard`.
 
+* `snapshot_id` - (Optional)  The ID of the Snapshot which should be used to create this cluster. Changing this forces a new resource to be created.
```
Comment (stephybun) [inline website/docs/r/kubernetes_cluster.html.markdown:238]: This appears to be a preview feature so can you please add a Note like we do for all the other preview features.

**Review by stephybun — CHANGES_REQUESTED** [review]
Also since this is a preview feature

**Comment by ms-henglu** [comment]
Hi @stephybun ,

Thank you for the code review! And apologizes that I didn't notice this feature is in preview, I'll close this PR until this feature is GA.

**Pull request closed without merging**